nntrainer/engine.h

```diff
@@ -0,0 +1,255 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2024 Jijoong Moon <jijoong.moon@samsung.com>
+ *
+ * @file   engine.h
+ * @date   27 December 2024
+ * @brief  This file contains engine context related functions and classes that
+ * manages the engines (NPU, GPU, CPU) of the current environment
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+
+#ifndef __ENGINE_H__
+#define __ENGINE_H__
+
+#include <algorithm>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <vector>
+
+#include <context.h>
+
+#include <nntrainer_error.h>
+
+namespace nntrainer {
+
+extern std::mutex engine_mutex;
+namespace {} // namespace
+
+/**
+ * @class Engine contains user-dependent configuration
+ * @brief App
+ */
+class Engine {
+protected:
+  static void registerer(Engine &eg) noexcept;
+
+  static void add_default_object(Engine &eg);
+
+  static void registerContext(std::string name, nntrainer::Context *context) {
+    const std::lock_guard<std::mutex> lock(engine_mutex);
+
+    std::transform(name.begin(), name.end(), name.begin(),
+                   [](unsigned char c) { return std::tolower(c); });
+
+    if (engines.find(name) != engines.end()) {
+      std::stringstream ss;
+      ss << "Cannot register Context with name : " << name;
+      throw std::invalid_argument(ss.str().c_str());
+    }
+    engines.insert(std::make_pair(name, context));
+  }
+
+public:
+  /**
+   * @brief   Default constructor
+   */
+  Engine() = default;
+
+  /**
+   * @brief   Default Destructor
+   */
+  ~Engine() = default;
+
+  /**
+   * @brief register a Context from a shared library
+   * plugin must have **extern "C" LayerPluggable *ml_train_context_pluggable**
+   * defined else error
+   *
+   * @param library_path a file name of the library
+   * @param base_path    base path to make a full path (optional)
+   * @throws std::invalid_parameter if library_path is invalid or library is
+   * invalid
+   */
+  int registerContext(const std::string &library_path,
+                      const std::string &base_path = "");
+
+  /**
+   * @brief get registered a Context
+   *
+   * @param name Registered Context Name
+   * @throws std::invalid_parameter if no context with name
+   * @return Context Pointer : for register Object factory, casting might be
+   * needed.
+   */
+  static nntrainer::Context *getRegisteredContext(std::string name) {
+
+    std::transform(name.begin(), name.end(), name.begin(),
+                   [](unsigned char c) { return std::tolower(c); });
+
+    if (engines.find(name) == engines.end()) {
+      throw std::invalid_argument("not registered");
+    }
+    return engines.at(name);
+  }
```

Comment:
Are we certain that we shouldn't take the mutex here?

It seems to me bit strange that mutator is protected by lock while `public static` observer of `engines` collection is not.